src/rt/tracegc.d

```diff
@@ -88,243 +84,147 @@ extern (C) void[] _d_newarrayT(const TypeInfo ti, size_t length);
-}
+    auto type_string = typeof(Declaration).stringof;
+    auto name = __traits(identifier, Declaration);
+    auto param_idx = findParamIndex(type_string);
 
-extern (C) byte[] _d_arrayappendcTXTrace(string file, int line, string funcname, const TypeInfo ti, ref byte[] px, size_t n)
-{
-    mixin(generatePrintf());
-    accumulate(file, line, funcname, ti.toString(), ti.next.tsize * n);
-    return _d_arrayappendcTX(ti, px, n);
-}
+    auto new_declaration = type_string[0 .. param_idx] ~ " " ~ name
+        ~ "Trace(string file, int line, string funcname, "
+        ~ type_string[param_idx+1 .. $];
+    auto call_original = "    scope original_func = { return "
+        ~ __traits(identifier, Declaration) ~ "(" ~ Arguments!Declaration() ~ "); };";
 
-extern (C) void[] _d_arrayappendcdTrace(string file, int line, string funcname, ref byte[] x, dchar c)
-{
-    mixin(generatePrintf());
-    size_t n;
-    if (c <= 0x7F)
-        n = 1;
-    else if (c <= 0x7FF)
-        n = 2;
-    else if (c <= 0xFFFF)
-        n = 3;
-    else if (c <= 0x10FFFF)
-        n = 4;
-    else
-        assert(0);
-    accumulate(file, line, funcname, "char[]", n * char.sizeof);
-    return _d_arrayappendcd(x, c);
+    return new_declaration ~ "\n{\n" ~
+           call_original ~ "\n" ~
+           accumulator ~ "\n" ~
+           "}\n";
 }
 
-extern (C) void[] _d_arrayappendwdTrace(string file, int line, string funcname, ref byte[] x, dchar c)
+string Arguments(alias Func)()
 {
-    mixin(generatePrintf());
-    size_t n = 1 + (c > 0xFFFF);
-    accumulate(file, line, funcname, "wchar[]", n * wchar.sizeof);
-    return _d_arrayappendwd(x, c);
-}
+    string result = "";
 
-extern (C) void[] _d_arraysetlengthT(const TypeInfo ti, size_t newlength, void[]* p);
-extern (C) void[] _d_arraysetlengthiT(const TypeInfo ti, size_t newlength, void[]* p);
-
-extern (C) void[] _d_arraysetlengthTTrace(string file, int line, string funcname, const TypeInfo ti, size_t newlength, void[]* p)
-{
-    mixin(generatePrintf());
-    accumulate(file, line, funcname, ti.toString(), ti.next.tsize * newlength);
-    return _d_arraysetlengthT(ti, newlength, p);
-}
+    static if (is(typeof(Func) PT == __parameters))
+    {
+        foreach (idx, _; PT)
+            result ~= __traits(identifier, PT[idx .. idx + 1]) ~ ", ";
+    }
 
-extern (C) void[] _d_arraysetlengthiTTrace(string file, int line, string funcname, const TypeInfo ti, size_t newlength, void[]* p)
-{
-    mixin(generatePrintf());
-    accumulate(file, line, funcname, ti.toString(), ti.next.tsize * newlength);
-    return _d_arraysetlengthiT(ti, newlength, p);
+    return result;
 }
 
-
-extern (C) void* _d_allocmemory(size_t sz);
-
-extern (C) void* _d_allocmemoryTrace(string file, int line, string funcname, size_t sz)
+unittest
 {
-    mixin(generatePrintf());
-    accumulate(file, line, funcname, "closure", sz);
-    return _d_allocmemory(sz);
+    void foo(int x, double y) { }
+    static assert (Arguments!foo == "x, y, ");
```

Comment:
OK, I got completely lost with all the code generation. I will let someone else more used to black magic to review it.